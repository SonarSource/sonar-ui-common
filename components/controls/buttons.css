/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  vertical-align: middle;
  height: var(--controlHeight);
  line-height: calc(var(--controlHeight) - 2px);
  padding: 0 var(--gridSize);
  border: 1px solid var(--darkBlue);
  border-radius: 2px;
  box-sizing: border-box;
  background: transparent;
  color: var(--darkBlue);
  font-weight: 600;
  font-size: var(--smallFontSize);
  text-decoration: none;
  cursor: pointer;
  outline: none;
  transition: border-color 0.2s ease, box-shadow 0.2s ease;
}

.button:hover,
.button.button-active {
  background: var(--darkBlue);
  color: #fff;
}

.button:active {
  box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.125);
}

.button:focus {
  box-shadow: 0 0 0 3px rgba(35, 106, 151, 0.25);
}

.button-primary {
  background: var(--darkBlue);
  color: #fff;
}

.button-primary:hover {
  background: var(--veryDarkBlue);
}

.button.disabled {
  color: var(--disableGrayText) !important;
  border-color: var(--disableGrayBorder) !important;
  background: var(--disableGrayBg) !important;
  cursor: not-allowed !important;
  box-shadow: none !important;
}

/* #region .button-red */
.button-red {
  border-color: var(--red);
  color: var(--red);
}

.button-red:hover,
.button-red.active {
  background: var(--red);
  color: #fff;
}

.button-red:focus {
  box-shadow: 0 0 0 3px rgba(212, 51, 63, 0.25);
}

/* #endregion */

/* #region .button-success */
.button-success {
  border-color: var(--green);
  color: var(--green);
}

.button-success:hover,
.button-success.active {
  background: var(--green);
  color: #fff;
}

.button-success:focus {
  box-shadow: 0 0 0 3px rgba(0, 170, 0, 0.25);
}

/* #endregion */

/* #region .button-link */
.button-link {
  display: inline-flex;
  height: auto;
  /* Keep this to not inherit the height from .button */
  line-height: 1;
  margin: 0;
  padding: 0;
  border: none;
  border-radius: 0;
  background: transparent;
  color: var(--darkBlue);
  border-bottom: 1px solid var(--lightBlue);
  font-weight: 400;
  font-size: inherit;
  transition: border-color 0.2s ease, box-shadow 0.2s ease, color 0.2s ease, border-bottom 0.2s ease;
}

.dropdown .button-link {
  border-bottom: none;
}

.button-link:hover {
  background: transparent;
  color: var(--blue);
}

.button-link:active,
.button-link:focus {
  box-shadow: none;
  outline: 1px dotted var(--blue);
}

.button-link.disabled {
  color: var(--secondFontColor);
  background: transparent !important;
  cursor: default;
}

/* #endregion */

.button-small {
  height: var(--smallControlHeight);
  line-height: 18px;
  padding: 0 6px;
  font-size: 11px;
}

.button-tiny {
  height: var(--tinyControlHeight);
  line-height: var(--tinyControlHeight);
  padding: 0 calc(var(--gridSize) / 2);
}

.button-large {
  height: var(--largeControlHeight);
  padding: 0 16px;
  font-size: var(--mediumFontSize);
}

.button-huge {
  flex-direction: column;
  padding: calc(2 * var(--gridSize));
  width: 200px;
  height: 200px;
  background-color: #fff;
  border: solid 1px #fff;
  border-radius: 3px;
  transition: all 0.2s ease;
  box-shadow: 0 1px 1px 1px var(--barBorderColor);
}

.button-huge:hover,
.button-huge:focus,
.button-huge:active {
  background-color: #fff;
  color: var(--darkBlue);
  box-shadow: var(--defaultShadow);
  transform: translateY(-2px);
}

/* #region .button-group */
/* TODO drop usage of this class in SQ (already dropped from SC) */
.button-group {
  display: inline-block;
  vertical-align: middle;
  font-size: 0;
  white-space: nowrap;
}

.button-group > button,
.button-group > .button {
  position: relative;
  z-index: var(--normalZIndex);
  display: inline-block;
  vertical-align: middle;
  margin: 0;
  cursor: pointer;
}

.button-group > .button:hover:not(.disabled),
.button-group > .button:focus:not(.disabled),
.button-group > .button:active:not(.disabled),
.button-group > .button.active:not(.disabled) {
  z-index: var(--aboveNormalZIndex);
}

.button-group > .button.disabled {
  z-index: var(--belowNormalZIndex);
}

.button-group > .button:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.button-group > .button:not(:last-child):not(.dropdown-toggle) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group > .button + .button {
  margin-left: -1px;
}

.button-group > a:not(.button) {
  vertical-align: middle;
  margin: 0 8px;
  font-size: var(--smallFontSize);
}

/* #endregion */

/* #region .button-icon */
.button-icon {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  vertical-align: middle;
  width: var(--controlHeight);
  height: var(--controlHeight);
  padding: 0;
  border: none;
  color: inherit;
}

.button-icon.button-small {
  width: var(--smallControlHeight);
  height: var(--smallControlHeight);
  padding: 0;
}

.button-icon.button-small svg {
  margin-top: 0;
}

.button-icon.button-tiny {
  width: var(--tinyControlHeight);
  height: var(--tinyControlHeight);
  padding: 0;
}

.button-icon.button-tiny svg {
  margin-top: 0;
}

.button-icon:hover,
.button-icon:focus {
  background-color: currentColor;
}

.button-icon:hover svg,
.button-icon:focus svg {
  color: #fff;
}

/* #endregion */

.button-list {
  display: inline-flex;
  justify-content: space-between;
  height: auto;
  border: 1px solid var(--barBorderColor);
  padding: var(--gridSize);
  margin: calc(var(--gridSize) / 2);
  color: var(--secondFontColor);
  font-weight: normal;
}

.button-list:hover {
  background-color: white;
  border-color: var(--blue);
  color: var(--darkBlue);
}

.no-select {
  user-select: none !important;
}
