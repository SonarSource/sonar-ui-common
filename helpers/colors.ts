/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
