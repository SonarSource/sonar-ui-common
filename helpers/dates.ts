/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
