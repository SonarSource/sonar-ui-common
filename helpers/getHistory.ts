/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createHistory, History } from 'history';
import { useRouterHistory } from 'react-router';

let history: History;

function ensureHistory() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  history = useRouterHistory(createHistory)({
    basename: (window as any).baseUrl,
  });
  return history;
}

export default function getHistory() {
  return history ? history : ensureHistory();
}
