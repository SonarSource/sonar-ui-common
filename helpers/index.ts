/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
