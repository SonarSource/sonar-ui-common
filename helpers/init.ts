/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import type { Messages } from './l10n';

let urlContext: string; // Is the base path (web context) in SQ
let messages: Messages | undefined;
let locale: string | undefined;

export const IS_SSR = typeof window === 'undefined';
export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.',
};
const LOGGER_PREFIX = 'sonar-ui-common init:';

export default {
  setUrlContext(newUrlContext: string) {
    urlContext = newUrlContext;
    return this;
  },
  setLocale(newLocale: string) {
    locale = newLocale;
    return this;
  },
  setMessages(newMessages: Messages) {
    messages = newMessages;
    return this;
  },
};

export function getMessages() {
  if (typeof messages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return messages;
}

export function getLocale() {
  if (typeof locale === 'undefined') {
    logWarning('L10n locale is not initialized. Use default locale.');
    return DEFAULT_LOCALE;
  }
  return locale;
}

export function getUrlContext() {
  if (typeof urlContext === 'undefined') {
    throw new Error(
      `${LOGGER_PREFIX} web context needs to be initialized by Initializer.setUrlContext before being used`
    );
  }
  return urlContext;
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(`${LOGGER_PREFIX} ${message}`);
  }
}
