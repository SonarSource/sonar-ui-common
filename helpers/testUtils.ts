/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
