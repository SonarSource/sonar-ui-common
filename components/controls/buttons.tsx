/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import ChevronRightIcon from '../icons/ChevronRightIcon';
import ClearIcon, { ClearIconProps } from '../icons/ClearIcon';
import DeleteIcon from '../icons/DeleteIcon';
import EditIcon from '../icons/EditIcon';
import { IconProps } from '../icons/Icon';
import { ThemeConsumer } from '../theme';
import './buttons.css';
import Tooltip, { TooltipProps } from './Tooltip';

type AllowedButtonAttributes = Pick<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  'className' | 'disabled' | 'id' | 'style' | 'title'
>;

interface ButtonProps extends AllowedButtonAttributes {
  autoFocus?: boolean;
  children?: React.ReactNode;
  innerRef?: (node: HTMLElement | null) => void;
  name?: string;
  onClick?: () => void;
  preventDefault?: boolean;
  stopPropagation?: boolean;
  type?: 'button' | 'submit' | 'reset' | undefined;
}

export class Button extends React.PureComponent<ButtonProps> {
  handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    const { disabled, onClick, preventDefault = true, stopPropagation = false } = this.props;

    event.currentTarget.blur();
    if (preventDefault) {
      event.preventDefault();
    }
    if (stopPropagation) {
      event.stopPropagation();
    }

    if (onClick && !disabled) {
      onClick();
    }
  };

  render() {
    const {
      className,
      disabled,
      innerRef,
      onClick,
      preventDefault,
      stopPropagation,
      type = 'button',
      ...props
    } = this.props;
    return (
      // eslint-disable-next-line react/button-has-type
      <button
        {...props}
        aria-disabled={disabled}
        className={classNames('button', className, { disabled })}
        id={this.props.id}
        onClick={this.handleClick}
        ref={this.props.innerRef}
        type={type}
      />
    );
  }
}

export function ButtonLink({ className, ...props }: ButtonProps) {
  return <Button {...props} className={classNames('button-link', className)} />;
}

export function SubmitButton(props: T.Omit<ButtonProps, 'type'>) {
  // do not prevent default to actually submit a form
  return <Button {...props} preventDefault={false} type="submit" />;
}

export function ResetButtonLink(props: T.Omit<ButtonProps, 'type'>) {
  return <ButtonLink {...props} type="reset" />;
}

export interface ButtonIconProps extends ButtonProps {
  'aria-label'?: string;
  'aria-labelledby'?: string;
  className?: string;
  color?: string;
  onClick?: () => void;
  tooltip?: React.ReactNode;
  tooltipProps?: Partial<TooltipProps>;
}

export function ButtonIcon(props: ButtonIconProps) {
  const { className, color, tooltip, tooltipProps, ...other } = props;
  return (
    <ThemeConsumer>
      {(theme) => (
        <Tooltip mouseEnterDelay={0.4} overlay={tooltip} {...tooltipProps}>
          <Button
            className={classNames(className, 'button-icon')}
            stopPropagation={true}
            style={{ color: color || theme.colors.darkBlue }}
            {...other}
          />
        </Tooltip>
      )}
    </ThemeConsumer>
  );
}

interface ClearButtonProps extends ButtonIconProps {
  className?: string;
  iconProps?: ClearIconProps;
  onClick?: () => void;
}

export function ClearButton({ color, iconProps = {}, ...props }: ClearButtonProps) {
  return (
    <ThemeConsumer>
      {(theme) => (
        <ButtonIcon color={color || theme.colors.gray60} {...props}>
          <ClearIcon {...iconProps} />
        </ButtonIcon>
      )}
    </ThemeConsumer>
  );
}

interface ActionButtonProps extends ButtonIconProps {
  className?: string;
  iconProps?: IconProps;
  onClick?: () => void;
}

export function DeleteButton({ iconProps = {}, ...props }: ActionButtonProps) {
  return (
    <ThemeConsumer>
      {(theme) => (
        <ButtonIcon color={theme.colors.red} {...props}>
          <DeleteIcon {...iconProps} />
        </ButtonIcon>
      )}
    </ThemeConsumer>
  );
}

export function EditButton({ iconProps = {}, ...props }: ActionButtonProps) {
  return (
    <ButtonIcon {...props}>
      <EditIcon {...iconProps} />
    </ButtonIcon>
  );
}

export function ListButton({ className, children, ...props }: ButtonProps) {
  return (
    <Button className={classNames('button-list', className)} {...props}>
      {children}
      <ChevronRightIcon />
    </Button>
  );
}
