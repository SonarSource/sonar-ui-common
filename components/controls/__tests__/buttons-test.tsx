/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { click, mockEvent } from '../../../helpers/testUtils';
import { Button, ButtonIcon, ButtonIconProps } from '../buttons';

describe('Button', () => {
  it('should render correctly', () => {
    const onClick = jest.fn();
    const preventDefault = jest.fn();
    const stopPropagation = jest.fn();
    const wrapper = shallowRender({ onClick });
    expect(wrapper).toMatchSnapshot();
    click(wrapper.find('button'), mockEvent({ preventDefault, stopPropagation }));
    expect(onClick).toBeCalled();
    expect(preventDefault).toBeCalled();
    expect(stopPropagation).not.toBeCalled();
  });

  it('should not stop propagation, but prevent default of the click event', () => {
    const preventDefault = jest.fn();
    const stopPropagation = jest.fn();
    const wrapper = shallowRender({ preventDefault: false, stopPropagation: true });
    click(wrapper.find('button'), mockEvent({ preventDefault, stopPropagation }));
    expect(preventDefault).not.toBeCalled();
    expect(stopPropagation).toBeCalled();
  });

  it('should disable buttons with a class', () => {
    const onClick = jest.fn();
    const button = shallowRender({ disabled: true, onClick }).find('button');
    expect(button.props().disabled).toBeUndefined();
    expect(button.props().className).toContain('disabled');
    expect(button.props()['aria-disabled']).toBe(true);
    click(button);
    expect(onClick).not.toBeCalled();
  });

  function shallowRender(props: Partial<Button['props']> = {}) {
    return shallow<Button>(<Button {...props}>My button</Button>);
  }
});

describe('ButtonIcon', () => {
  it('should render correctly', () => {
    const wrapper = shallowRender();
    expect(wrapper).toMatchSnapshot();
  });

  function shallowRender(props: Partial<ButtonIconProps> = {}) {
    return shallow(
      <ButtonIcon tooltip="my tooltip" tooltipProps={{ visible: true }} {...props}>
        <i />
      </ButtonIcon>
    ).dive();
  }
});
