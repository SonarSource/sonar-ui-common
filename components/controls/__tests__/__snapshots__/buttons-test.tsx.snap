// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<button
  className="button"
  onClick={[Function]}
  type="button"
>
  My button
</button>
`;

exports[`ButtonIcon should render correctly 1`] = `
<Tooltip
  mouseEnterDelay={0.4}
  overlay="my tooltip"
  visible={true}
>
  <Button
    className="button-icon"
    stopPropagation={true}
    style={
      Object {
        "color": "#236a97",
      }
    }
  >
    <i />
  </Button>
</Tooltip>
`;
